monetio/sat/_tropomi_l2_no2_mm.py

```diff
@@ -0,0 +1,138 @@
+"""Read TROPOMI L2 NO2 data."""
+
+import logging
+import os
+import sys
+from collections import OrderedDict
+from glob import glob
+
+import numpy as np
+import xarray as xr
+from netCDF4 import Dataset
+
+
+def _open_one_dataset(fname, variable_dict):
+    """
+    Parameters
+    ----------
+    fname : str
+        Input file path.
+    variable_dict : dict
+
+    Returns
+    -------
+    xarray.Dataset
+    """
+    print("reading " + fname)
+
+    ds = xr.Dataset()
+
+    dso = Dataset(fname, "r")
+
+    longitude = dso.groups["PRODUCT"]["longitude"]
+    latitude = dso.groups["PRODUCT"]["latitude"]
+    start_time = dso.groups["PRODUCT"]["time"]
+
+    # squeeze 1-dimension
+    longitude = np.squeeze(longitude)
+    latitude = np.squeeze(latitude)
+    start_time = np.squeeze(start_time)
+
+    ds["lon"] = xr.DataArray(longitude)
+    ds["lat"] = xr.DataArray(latitude)
+
+    for varname in variable_dict:
+        print(varname)
+        values = dso.groups["PRODUCT"][varname]
+        # squeeze out 1-dimension
+        values = np.squeeze(values)
+
+        if "fillvalue" in variable_dict[varname]:
+            fillvalue = variable_dict[varname]["fillvalue"]
+            values[:][values[:] == fillvalue] = np.nan
+
+        if "scale" in variable_dict[varname]:
+            values[:] = variable_dict[varname]["scale"] * values[:]
+
+        if "minimum" in variable_dict[varname]:
+            minimum = variable_dict[varname]["minimum"]
+            values[:][values[:] < minimum] = np.nan
+
+        if "maximum" in variable_dict[varname]:
+            maximum = variable_dict[varname]["maximum"]
+            values[:][values[:] > maximum] = np.nan
+
+        ds[varname] = xr.DataArray(values)
+
+        if "quality_flag_min" in variable_dict[varname]:
+            ds.attrs["quality_flag"] = varname
+            ds.attrs["quality_thresh_min"] = variable_dict[varname]["quality_flag_min"]
+
+    dso.close()
+
+    return ds
+
+
+def apply_quality_flag(ds):
+    """Mask variables in place based on quality flag.
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+    """
+    if "quality_flag" in ds.attrs:
+        quality_flag = ds[ds.attrs["quality_flag"]]
+        quality_thresh_min = ds.attrs["quality_thresh_min"]
+
+        # Apply the quality thresh minimum to all variables in ds
+        for varname in ds:
+            print(varname)
+            if varname != ds.attrs["quality_flag"]:
+                logging.debug(varname)
+                values = ds[varname].values
+                values[quality_flag <= quality_thresh_min] = np.nan
+
+
+def open_dataset(fnames, variable_dict, debug=False):
```

Comment:
The `variable_dict` is necessary for the fn to be useful. Is this something that will always be the same? If so, could include in the code sort of like in the MOPITT one.

and/or document in the docstring the format

Reply:
fn always change with the date and vary one by one. It seems that there's no string always the same.

Reply:
By the way, by "fn" I meant function, not filename. I was wondering if we could set it up so you could pass `varnames` like the MOPITT one, instead of the variable dict. The variable dict info could be embedded in the module, if it is fixed for a certain variable.